### PR TITLE
[ROM] Complement offline stage of ROM

### PR DESCRIPTION
**Pull request by Rbravo555**
#depends on #9503

**📝 Description**
This PR adds the possibility to modify the parameters of a ROM simulation by editing the RomParameters.json.

In Particular, to indicate that a ROM simulation should be used for training the HROM elements and weights, it was necessary to pass a string to RomAnalysis. Now, one needs to set  "train_rom": True in the RomParameters.json.

**🆕 Changelog**
- Getting rid of the ElementsAndWeights.json file. Instead, its content is appended to the RomParameters.json
- Getting rid of the CreateHRomAnalysisInstance of the rom_analysis
- Modifying parameters of the rom_output_process to include train_hrom and run_hrom

### CODE REVIEW

**Comment by rubenzorrilla** [comment]
As the previous implementation was way intrusive, I created the `AuxiliaryRomUtilities`. In there I've defined two static methods, one for the creation of the HROM computing model part and the visualization one. Regarding the former, I've basically refactored the previous Python code to c++. Concerning the latter, I think that the solution we had so far is not very convenient, as it relies on the existent conditions to create the visualization mesh. Chances are that there are no conditions, so no visualization mesh is created. Instead, I suggest to detect the skin and create those entities manually (in this PR I'm only adding the interface, I'd add this capability in a different PR).

Another open issue is to add the capability to optionally add the condition together with its parent element in the HROM computing model part creation as this might be needed in some formulations. 

(I didn't test the implementation yet)

**Comment by rubenzorrilla** [comment]
@Rbravo555 @SADPR I think we should start thinking about merging this (otherwise the PR will get bigger and bigger). On top of the HROM workflow enhancements, now we have the capability to add HROM condition parents and minimum conditions. Besides this, the skin of a volumetric body can be generated as a visualization mesh. To this regard, in a future PR we should create a process to do the solution projection onto this mesh (as well as the BC imposition) and the output. Note that this comes after the fact that the **visualization model parts are independent to the HROM one**.

**Comment by rubenzorrilla** [comment]
#9608 must be merged first.

File: applications/RomApplication/custom_utilities/rom_auxiliary_utilities.cpp [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+    NodesPointerSetType hrom_nodes_set;
+    std::vector<Element::Pointer> hrom_elems_vect;
+    std::vector<Condition::Pointer> hrom_conds_vect;
+
+    const auto& r_elem_weights = HRomWeights["Elements"];
+    hrom_elems_vect.reserve(rOriginModelPart.NumberOfElements());
+    for (auto it = r_elem_weights.begin(); it != r_elem_weights.end(); ++it) {
+        // Get element from origin mesh
+        const IndexType elem_id = stoi(it.name());
+        const auto p_elem = rOriginModelPart.pGetElement(elem_id + 1); //FIXME: WHY THIS +1?
```
Comment (SADPR) [inline applications/RomApplication/custom_utilities/rom_auxiliary_utilities.cpp:62]: Is it going to remain +1?  For me, it is ok for now...

Comment (rubenzorrilla) [reply applications/RomApplication/custom_utilities/rom_auxiliary_utilities.cpp:62]: I'd change the Ids 0 issue in a standalone PR to easily backtrack the changes if something gets messed up.

File: applications/RomApplication/python_scripts/rom_analysis.py [hunk 2]
```diff
@@ -46,78 +86,67 @@ def ModifyAfterSolverInitialize(self):
...
-                    res_mat = self.ResidualUtilityObject.GetResiduals()
-                    np_res_mat = np.array(res_mat, copy=False)
-                    self.time_step_residual_matrix_container.append(np_res_mat)
-
+            # Call the HROM training utility to append the current step residuals
+            # Note that this needs to be done prior to the other processes to avoid unfixing the BCs
+            if self.train_hrom:
+                self.__hrom_training_utility.AppendCurrentStepResiduals()
+
+            # #FIXME: Make this optional. This must be a process
```
Comment (SADPR) [inline applications/RomApplication/python_scripts/rom_analysis.py:117]: I'm confused with this, isn't it always projected in ProjectToFineBasis?

Comment (rubenzorrilla) [reply applications/RomApplication/python_scripts/rom_analysis.py:117]: The point is that the visualization model parts must be independent to the HROM one. I'll take care of this in an upcoming PR. So far I only included some tools required for the creation of such model parts.

**Review by SADPR — COMMENTED** [review]
It looks good to me. @Rbravo555 check it so we can merge it.

File: applications/RomApplication/python_scripts/rom_analysis.py [hunk 3]
```diff
@@ -4,6 +4,7 @@
 import KratosMultiphysics.RomApplication as KratosROM
 from KratosMultiphysics.RomApplication import new_python_solvers_wrapper_rom
 from KratosMultiphysics.RomApplication.hrom_training_utility import HRomTrainingUtility
+from KratosMultiphysics.RomApplication.calculate_rom_basis_output_process import CalculateRomBasisOutputProcess
```
Comment (rubenzorrilla) [inline applications/RomApplication/python_scripts/rom_analysis.py:7]: Why is this required? The process must be imported by the factory when instantiating the content of the `output_process_list`?

Comment (Rbravo555) [reply applications/RomApplication/python_scripts/rom_analysis.py:7]: I noticed the ROM simulation runs well right after the training. But when launching it separately (for changing the number of modes to use for example) Kratos complained that the “calculate_rom_output_process” didn’t belong to the RomApplication. Importing it manually here solved the problem for me..

Comment (rubenzorrilla) [reply applications/RomApplication/python_scripts/rom_analysis.py:7]: OK. My bet is that if the process is not added in the process list it is not imported, then it cannot do the check below.

**Review by rubenzorrilla — APPROVED** [review]
